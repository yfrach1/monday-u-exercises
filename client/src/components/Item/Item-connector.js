import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import {
  deleteItemByIdAction,
  toggleStatusAction,
  updateItemNameAction,
} from "../../app/actions/items-entities-actions";
import Item from "./Item";

const mapStateToProps = (state, ownProps) => {
  // const items = getItems(state);
  return {};
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return bindActionCreators(
    { deleteItemByIdAction, toggleStatusAction, updateItemNameAction },
    dispatch
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(Item);
