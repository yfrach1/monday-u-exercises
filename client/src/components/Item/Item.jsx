import React, { useCallback, useState } from "react";
import styles from "./Item.module.css";
import deleteIcon from "../../assets/images/delete_icon.svg";
import editIcon from "../../assets/images/edit_icon.svg";
import saveIcon from "../../assets/images/save_icon.svg";
import PropTypes from "prop-types";
import noop from "react-props-noop";
import {
  deleteTaskById,
  flipStatus,
  updateTaskText,
} from "../../serverApi/itemClient";
import Icon from "../Icon/Icon";

const Item = ({
  itemName,
  id,
  status,
  removeIdFromDataHandler,
  onChangeValueUpdateDataHandler,
  checkIfTextAlreadyExist,
  displayToast,
  setShowToast,
  killToast,
}) => {
  const [showSaveButton, setShowSaveButton] = useState(false);
  const [showEditButton, setShowEditButton] = useState(true);
  const [textReadOnly, setTextReadOnly] = useState(true);
  const [isEdit, setIsEdit] = useState(false);
  const [textAfterEdit, setTextAfterEdit] = useState(itemName);

  const onClickEditItemHandler = () => {
    setShowSaveButton(true);
    setShowEditButton(false);
    setTextReadOnly(false);
    setIsEdit(true);
  };

  const onClickSaveItemHandler = async () => {
    setShowSaveButton(false);
    setShowEditButton(true);
    setTextReadOnly(true);
    const textBeforeEdit = itemName;
    const editTaskResult = await updateTaskText(id, textAfterEdit);
    const isNewTextAlreadyExist = checkIfTextAlreadyExist(editTaskResult.data);
    if (editTaskResult.result === "success") {
      if (isNewTextAlreadyExist) {
        if (textBeforeEdit !== textAfterEdit) {
          displayToast("", "Task already exist in ToDO App");
        }
      } else {
        setTextAfterEdit(editTaskResult.data);
        onChangeValueUpdateDataHandler(id, "itemName", editTaskResult.data);
        displayToast("POSITIVE", "Task successfully edited & saved");

        killToast();
      }
    } else {
      displayToast("NEGATIVE", editTaskResult.data);
    }

    setIsEdit(false);
  };

  const onClickDeleteItemHandler = useCallback(async () => {
    const deleteByIdResult = await deleteTaskById(id);
    if (deleteByIdResult.result === "success") {
      removeIdFromDataHandler(id);
      displayToast("POSITIVE", "Task successfully deleted");
      killToast();
    } else {
      displayToast("NEGATIVE", deleteByIdResult.data);
    }
  }, [id, removeIdFromDataHandler, setShowToast, displayToast]);

  const onClickstatusCheckBoxHandler = useCallback(async () => {
    const flipStatusResult = await flipStatus(id);
    if (flipStatusResult.result === "success") {
      onChangeValueUpdateDataHandler(id, "status", flipStatusResult.data);
      const messaeg = flipStatusResult.data
        ? "Task status successfully checked"
        : "Task status successfully unchecked";
      displayToast("POSITIVE", messaeg);

      killToast();
    } else {
      displayToast("NEGATIVE", flipStatusResult.data);
    }
  }, [id, onChangeValueUpdateDataHandler, setShowToast, displayToast]);

  const onChangeTaskTextHandler = (e) => {
    setTextAfterEdit(e.target.value);
  };
  return (
    <div
      className={`${styles.listItem} ${status ? styles.inputTextTaskDone : ""}`}
    >
      <input
        className={styles.itemStatus}
        type="checkbox"
        checked={status}
        onChange={onClickstatusCheckBoxHandler}
      />

      <input
        className={isEdit ? styles.itemTextEditMode : styles.itemText}
        type="text"
        value={!isEdit ? itemName : textAfterEdit}
        readOnly={textReadOnly}
        onChange={onChangeTaskTextHandler}
      />
      {showEditButton && !status && (
        <Icon
          iconClassName={"listItemIcon"}
          style={{ marginLeft: "auto" }}
          src={editIcon}
          onClickIconHandler={onClickEditItemHandler}
        />
      )}
      {showSaveButton && !status && (
        <Icon
          iconClassName={"listItemIcon"}
          style={{ marginLeft: "auto" }}
          src={saveIcon}
          onClickIconHandler={onClickSaveItemHandler}
        />
      )}
      <Icon
        iconClassName={"listItemIcon"}
        style={status ? { marginLeft: "auto" } : {}}
        src={deleteIcon}
        onClickIconHandler={onClickDeleteItemHandler}
      />
    </div>
  );
};

Item.propTypes = {
  itemName: PropTypes.string,
  id: PropTypes.number,
  status: PropTypes.bool,
  removeIdFromDataHandler: PropTypes.func,
  onChangeValueUpdateDataHandler: PropTypes.func,
  checkIfTextAlreadyExist: PropTypes.func,
  displayToast: PropTypes.func,
  setShowToast: PropTypes.func,
  killToast: PropTypes.func,
};

Item.defaultProps = {
  itemName: "task text",
  id: 0,
  status: false,
  removeIdFromDataHandler: noop,
  onChangeValueUpdateDataHandler: noop,
  checkIfTextAlreadyExist: noop,
  displayToast: noop,
  setShowToast: noop,
  killToast: noop,
};

export default Item;
