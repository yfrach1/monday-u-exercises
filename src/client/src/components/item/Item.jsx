import React, { useCallback, useState } from "react";
import styles from "./Item.module.css";
import deleteIcon from "../../assets/images/delete_icon.svg";
import editIcon from "../../assets/images/edit_icon.svg";
import saveIcon from "../../assets/images/save_icon.svg";
import PropTypes from "prop-types";
import Icon from "../Icon/Icon";
import { capitalizeFirstLetter } from "../../utils/utils";

const Item = ({
  id,
  itemName,
  status,
  deleteItemByIdAction,
  toggleStatusAction,
  updateItemNameAction,
  checkIfTextAlreadyExist,
}) => {
  const [showSaveButton, setShowSaveButton] = useState(false);
  const [showEditButton, setShowEditButton] = useState(true);
  const [textReadOnly, setTextReadOnly] = useState(true);
  const [isEdit, setIsEdit] = useState(false);
  const [textAfterEdit, setTextAfterEdit] = useState(itemName);
  const onClickEditItemHandler = () => {
    setShowSaveButton(true);
    setShowEditButton(false);
    setTextReadOnly(false);
    setIsEdit(true);
  };

  const onClickSaveItemHandler = async () => {
    setShowSaveButton(false);
    setShowEditButton(true);
    setTextReadOnly(true);
    console.log(!checkIfTextAlreadyExist(capitalizeFirstLetter(textAfterEdit)));
    if (!checkIfTextAlreadyExist(capitalizeFirstLetter(textAfterEdit))) {
      await updateItemNameAction(id, textAfterEdit, itemName);
    } else {
      setTextAfterEdit(itemName);
    }
    setIsEdit(false);
  };

  const onClickDeleteItemHandler = useCallback(async () => {
    await deleteItemByIdAction(id);
  }, [id, deleteItemByIdAction]);

  const onClickstatusCheckBoxHandler = useCallback(async () => {
    await toggleStatusAction(id);
  }, [id, toggleStatusAction]);

  const onChangeTaskTextHandler = (e) => {
    setTextAfterEdit(e.target.value);
  };
  return (
    <li
      className={`${styles.listItem} ${status ? styles.inputTextTaskDone : ""}`}
    >
      <input
        className={styles.itemStatus}
        type="checkbox"
        checked={status}
        onChange={onClickstatusCheckBoxHandler}
      />

      <input
        className={isEdit ? styles.itemTextEditMode : styles.itemText}
        type="text"
        value={!isEdit ? itemName : textAfterEdit}
        readOnly={textReadOnly}
        onChange={onChangeTaskTextHandler}
      />
      {showEditButton && !status && (
        <Icon
          iconClassName={"listItemIcon"}
          style={{ marginLeft: "auto" }}
          src={editIcon}
          onClickIconHandler={onClickEditItemHandler}
        />
      )}
      {showSaveButton && !status && (
        <Icon
          iconClassName={"listItemIcon"}
          style={{ marginLeft: "auto" }}
          src={saveIcon}
          onClickIconHandler={onClickSaveItemHandler}
        />
      )}
      <Icon
        iconClassName={"listItemIcon"}
        id={"delete_icon"}
        style={status ? { marginLeft: "auto" } : {}}
        src={deleteIcon}
        onClickIconHandler={onClickDeleteItemHandler}
      />
    </li>
  );
};

Item.propTypes = {
  itemName: PropTypes.string,
  id: PropTypes.number,
  status: PropTypes.bool,
  deleteItemByIdAction: PropTypes.func,
  toggleStatusAction: PropTypes.func,
  updateItemNameAction: PropTypes.func,
  checkIfTextAlreadyExist: PropTypes.func,
};

Item.defaultProps = {
  itemName: "itemName",
  id: 0,
  status: false,
  deleteItemByIdAction: () => {},
  toggleStatusAction: () => {},
  updateItemNameAction: () => {},
  checkIfTextAlreadyExist: () => {},
};

export default Item;
