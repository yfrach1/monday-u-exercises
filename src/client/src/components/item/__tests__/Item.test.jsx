import React from "react";
import renderer from "react-test-renderer";
import Item from "../Item";

describe("Item Snapshot ", () => {
  const itemName = "itemName example";
  const id = 99999;
  const status = false;
  const checkIfTextAlreadyExist = (taskText) => {
    taskText ? true : false;
  };

  test("renders Item component with no variable sending via props from parent component", () => {
    const tree = renderer.create(<Item />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders Item component with no variable sending via props from parent component", () => {
    const tree = renderer
      .create(
        <Item
          itemName={itemName}
          id={id}
          status={status}
          checkIfTextAlreadyExist={checkIfTextAlreadyExist}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders Item component without itemName ", () => {
    const tree = renderer
      .create(
        <Item
          id={id}
          status={true} // check it
          checkIfTextAlreadyExist={checkIfTextAlreadyExist}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders Item component without checkIfTextAlreadyExist function ", () => {
    const tree = renderer
      .create(<Item itemName={itemName} status={status} id={id} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
