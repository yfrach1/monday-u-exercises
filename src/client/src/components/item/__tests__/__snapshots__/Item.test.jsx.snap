// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Snapshot  renders Item component with no variable sending via props from parent component 1`] = `
<li
  className="listItem "
>
  <input
    checked={false}
    className="itemStatus"
    onChange={[Function]}
    type="checkbox"
  />
  <input
    className="itemText"
    onChange={[Function]}
    readOnly={true}
    type="text"
    value="itemName"
  />
  <img
    alt=""
    className="listItemIcon"
    onClick={[Function]}
    src="edit_icon.svg"
    style={
      Object {
        "marginLeft": "auto",
      }
    }
  />
  <img
    alt=""
    className="listItemIcon"
    onClick={[Function]}
    src="delete_icon.svg"
    style={Object {}}
  />
</li>
`;

exports[`Item Snapshot  renders Item component with no variable sending via props from parent component 2`] = `
<li
  className="listItem "
>
  <input
    checked={false}
    className="itemStatus"
    onChange={[Function]}
    type="checkbox"
  />
  <input
    className="itemText"
    onChange={[Function]}
    readOnly={true}
    type="text"
    value="itemName example"
  />
  <img
    alt=""
    className="listItemIcon"
    onClick={[Function]}
    src="edit_icon.svg"
    style={
      Object {
        "marginLeft": "auto",
      }
    }
  />
  <img
    alt=""
    className="listItemIcon"
    onClick={[Function]}
    src="delete_icon.svg"
    style={Object {}}
  />
</li>
`;

exports[`Item Snapshot  renders Item component without checkIfTextAlreadyExist function  1`] = `
<li
  className="listItem "
>
  <input
    checked={false}
    className="itemStatus"
    onChange={[Function]}
    type="checkbox"
  />
  <input
    className="itemText"
    onChange={[Function]}
    readOnly={true}
    type="text"
    value="itemName example"
  />
  <img
    alt=""
    className="listItemIcon"
    onClick={[Function]}
    src="edit_icon.svg"
    style={
      Object {
        "marginLeft": "auto",
      }
    }
  />
  <img
    alt=""
    className="listItemIcon"
    onClick={[Function]}
    src="delete_icon.svg"
    style={Object {}}
  />
</li>
`;

exports[`Item Snapshot  renders Item component without itemName  1`] = `
<li
  className="listItem inputTextTaskDone"
>
  <input
    checked={true}
    className="itemStatus"
    onChange={[Function]}
    type="checkbox"
  />
  <input
    className="itemText"
    onChange={[Function]}
    readOnly={true}
    type="text"
    value="itemName"
  />
  <img
    alt=""
    className="listItemIcon"
    onClick={[Function]}
    src="delete_icon.svg"
    style={
      Object {
        "marginLeft": "auto",
      }
    }
  />
</li>
`;
